src/binary/timestamp.rs

```diff
@@ -67,13 +73,25 @@ where
                         VarUInt::write_u64(self, utc.second() as u64)?;
                     }
                     if timestamp.precision > Precision::Second {
-                        // This implementation always writes the datetime out with nanosecond precision.
-                        // The nanoseconds are an integer, which means that our exponent is always 9
-                        // and our coefficient is the number of nanoseconds.
-                        let nanoseconds = utc.nanosecond();
-                        const SCALE: i64 = -9; // "Scale" (used by BigDecimal) is -1 * exponent
-                        VarInt::write_i64(self, SCALE)?;
-                        DecodedUInt::write_u64(self, nanoseconds as u64)?;
+                        // It's not possible to have a precision of fractional
+                        // seconds and then not have any!
```

Comment:
Consider turning this into an [`unreachable!`](https://doc.rust-lang.org/std/macro.unreachable.html):
```
else {
  unreachable!("timestamps with fractional second precision must have fractional seconds");
}
```

Reply:
I think we want to have the property that this crate _never_ panics. So even though this code is actually unreachable, I don't want to make our auditing difficult. I was thinking about adding a new `IonError` variant for non-panic unreachable, wdyt?